ibis/expr/format.py

```diff
@@ -1,283 +1,598 @@
+    cols = [f"{name:<{maxlen}} {typ}" for name, typ in schema.items()]
+    depth = ibis.options.repr.table_columns
+    if depth is not None and depth < len(cols):
+        first_column_name = names[0]
+        raw = fmt_truncated(
+            cols,
+            depth=depth,
+            sep="\n",
+            ellipsis=util.VERTICAL_ELLIPSIS.center(len(first_column_name)),
+        )
+    else:
+        raw = "\n".join(cols)
+
+    return util.indent(raw, spaces=2)
+
+
+@fmt_table_op.register
+def _(op: ops.SQLQueryResult, **_: Any) -> str:
+    short_query = textwrap.shorten(
+        op.query,
+        ibis.options.repr.query_text_length,
+        placeholder=f" {util.HORIZONTAL_ELLIPSIS}",
+    )
+    query = f"query: {short_query!r}"
+    top = op.__class__.__name__
+    formatted_schema = fmt_schema(op.schema)
+    schema_field = util.indent(f"schema:\n{formatted_schema}", spaces=2)
+    return f"{top}\n{util.indent(query, spaces=2)}\n{schema_field}"
+
+
+@functools.singledispatch
+def fmt_join(op: ops.Join, *, aliases: Aliases) -> tuple[str, str]:
+    assert False, f"join type {type(op)} not implemented"
+
+
+@fmt_join.register(ops.Join)
+def _(op: ops.Join, *, aliases: Aliases) -> tuple[str, str]:
+    # format the operator and its relation inputs
+    left = aliases[op.left.op()]
+    right = aliases[op.right.op()]
+    top = f"{op.__class__.__name__}[{left}, {right}]"
+
+    # format the join predicates
+    # if only one, put it directly after the join on thes same line
+    # if more than one put each on a separate line
+    preds = op.predicates
+    formatted_preds = [fmt_value(pred, aliases=aliases) for pred in preds]
+    has_one_pred = len(preds) == 1
+    sep = " " if has_one_pred else "\n"
+    joined_predicates = util.indent(
+        "\n".join(formatted_preds),
+        spaces=2 * (not has_one_pred),
+    )
+    trailing_sep = "\n" + "\n" * (not has_one_pred)
+    return f"{top}{sep}{joined_predicates}", trailing_sep
+
+
+@fmt_join.register(ops.AsOfJoin)
+def _(op: ops.AsOfJoin, *, aliases: Aliases) -> tuple[str, str]:
+    left = aliases[op.left.op()]
+    right = aliases[op.right.op()]
+    top = f"{op.__class__.__name__}[{left}, {right}]"
+    raw_parts = fmt_fields(
+        op,
+        dict(predicates=fmt_value, by=fmt_value, tolerance=fmt_value),
+        aliases=aliases,
+    )
+    return f"{top}\n{raw_parts}", "\n\n"
+
+
+@fmt_table_op.register
+def _(
+    op: ops.Join,
+    *,
+    aliases: Aliases,
+    deps: Deque[ops.TableNode],
+    alias: Alias,
+    **_: Any,
+) -> str:
+    # first, format the current join operation
+    result, join_sep = fmt_join(op, aliases=aliases)
+    formatted_joins = [result, join_sep]
+
+    # process until the first non-Join dependency is popped in other words
+    # process all runs of joins
+    current = None
+    while deps and isinstance(current := deps.popleft(), ops.Join):
+        # bump the current node's alias
+        alias.value += 1
```

Comment:
Could we use an immutable alias instead? 
Could we traverse the operation dag in advance and locate the unique table nodes, assign aliases and lookup during the formatting?

Reply:
Lookup cannot be done in advance without big changes. I can certainly assign aliases up front though.